### PR TITLE
fix romimg bundled on docker containers

### PR DESCRIPTION
**Pull request by israpps**
instead of relying on strlen, go straight away with max bufsize. before this change the romimg from docker would segfault while local build works like a charm

New feature, `-c` and `-a` commands will have new variants, `-C` and `-A`, these will manipulate the filename written to ROMFS removing extension and converting to uppercase

So for example
```bash
romimg -C IOPRP.IMG $PS2SDK/iop/irx/secrman.irx
```
Will register secrman.irx as `SECRMAN` on the image, this seems to be needed when the IOPRP is replacing startup IOP modules

### CODE REVIEW

File: tools/romimg/src/romimg.c [hunk 1]
```diff
@@ -154,13 +150,15 @@ int CreateBlankROMImg(const char *filename, ROMIMG *ROMImg)
...
 #else
-	UserName = getenv("USER");
+    getlogin_r(UserName, sizeof(UserName));
 #endif
 	GetLocalhostName(LocalhostName, sizeof(LocalhostName));
 	GetCurrentWorkingDirectory(cwd, sizeof(cwd));
 	/* Comment format: YYYYMMDD-XXXYYY,conffile,<filename>,<user>@<localhost>/<image path> */
-	CommentLength = 31 + strlen(filename) + strlen(UserName) + strlen(LocalhostName) + strlen(cwd);
-	ROMImg->comment = (char *)malloc(CommentLength);
+	CommentLength = 31 + strlen(filename) + 32 + 32 + 128;
```
[the remark this answers is not in the document]
Comment (fjtrujy) [reply tools/romimg/src/romimg.c:158]: Another alternative is to use `sizeof(LocalhostName)`, this should return 32 as well

File: tools/romimg/src/romimg.c [hunk 2]
```diff
@@ -139,12 +140,7 @@ static int GetExtInfoStat(const struct ROMImgStat *ImageStat, struct RomDirFileF
 int CreateBlankROMImg(const char *filename, ROMIMG *ROMImg)
 {
 	unsigned int CommentLength;
-	char LocalhostName[32], cwd[128];
-#if defined(_WIN32) || defined(WIN32)
-	char UserName[32] = "";
-#else
-	char* UserName;
-#endif
+	char LocalhostName[32] = "\0", cwd[128] = "\0", UserName[32] = "\0";
```
[the remark this answers is not in the document]
Comment (israpps) [reply tools/romimg/src/romimg.c]: I guess there is no cross platform definition of MAX_PATH. linux seems to use PATH_MAX

[the remark this answers is not in the document]
Comment (fjtrujy) [reply tools/romimg/src/romimg.c]: Yeah sorry PATH_MAX or whatever is called, I don't remember right now the POSIX solution for that hehe

**Comment by uyjulian** [comment]
Add or remove one from either buffer or length

**Comment by israpps** [comment]
> Add or remove one from either buffer or length

will take a look. buffer is 11 and i'm reading 10 already

**Comment by uyjulian** [comment]
The buffer is being truncated.

That is, the destination buffer is smaller than the source buffer. So increase the destination buffer size or decrease the source buffer size

**Comment by israpps** [comment]
im kinda lost now
```
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10: error: ‘__builtin_strncpy’ output may be truncated copying 9 bytes from a string of length 9 [-Werror=stringop-truncation]
```

File: tools/romimg/src/romimg.c [hunk 3]
```diff
@@ -6,11 +6,16 @@
...
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
+#include <limits.h>
 
 #include "platform.h"
 #include "romimg.h"
 #include "SonyRX.h"
 
+#define BUFCHK(X) (X[0] == '\0') ? "" : X
```
Comment (fjtrujy) [inline tools/romimg/src/romimg.c:16]: Do you think this is really needed? 
I think it makes no sense at all, the behavior should be the same, no matter if the first character is `\0`

Comment (israpps) [reply tools/romimg/src/romimg.c:16]: yeah, it remained from older approach. will deal with it now

**Comment by israpps** [comment]
fixed. seems to work as expected!

**Comment by fjtrujy** [comment]
Finally 😄

File: tools/romimg/src/romimg.c [hunk 4]
```diff
@@ -154,14 +153,15 @@ int CreateBlankROMImg(const char *filename, ROMIMG *ROMImg)
...
-	UserName = getenv("USER");
+    getlogin_r(UserName, sizeof(UserName));
 #endif
 	GetLocalhostName(LocalhostName, sizeof(LocalhostName));
 	GetCurrentWorkingDirectory(cwd, sizeof(cwd));
 	/* Comment format: YYYYMMDD-XXXYYY,conffile,<filename>,<user>@<localhost>/<image path> */
-	CommentLength = 31 + strlen(filename) + strlen(UserName) + strlen(LocalhostName) + strlen(cwd);
-	ROMImg->comment = (char *)malloc(CommentLength);
-	sprintf(ROMImg->comment, "%08x,conffile,%s,%s@%s/%s", ROMImg->date, filename, (UserName[0] =='\0')?"":UserName, LocalhostName, cwd);
+	CommentLength = IMAGE_COMMENT_BASESIZE + strlen(filename) + sizeof(LocalhostName) + sizeof(UserName) + MAX_PATH;
```
Comment (fjtrujy) [inline tools/romimg/src/romimg.c:161]: use `sizeof` at the end

File: tools/romimg/src/platform.h [hunk 5]
```diff
@@ -6,11 +6,15 @@ int GetLocalhostName(char *buffer, unsigned int BufferSize);
...
 int GetCurrentWorkingDirectory(char *buffer, unsigned int BufferSize);
+void upperbuff(char *temp);
 
 #if defined(_WIN32) || defined(WIN32)
 #define PATHSEP '\\'
 #else
 #define PATHSEP '/'
 #endif
+#ifndef MAX_PATH
+#define MAX_PATH 4096
```
Comment (fjtrujy) [inline tools/romimg/src/platform.h:17]: `PATH_MAX` is available in all the OS, so use it instead and remove this from here

File: tools/romimg/src/romimg.c [hunk 6]
```diff
@@ -139,12 +143,7 @@ static int GetExtInfoStat(const struct ROMImgStat *ImageStat, struct RomDirFileF
 int CreateBlankROMImg(const char *filename, ROMIMG *ROMImg)
 {
 	unsigned int CommentLength;
-	char LocalhostName[32], cwd[128];
-#if defined(_WIN32) || defined(WIN32)
-	char UserName[32] = "";
-#else
-	char* UserName;
-#endif
+	char LocalhostName[32] = "\0", cwd[MAX_PATH] = "\0", UserName[32] = "\0";
```
Comment (fjtrujy) [inline tools/romimg/src/romimg.c:146]: I recommend to initializate them with `= {0}` instead, it is equivalent to put all the fields with `\0` directly

File: tools/romimg/src/romimg.c [hunk 7]
```diff
@@ -452,15 +452,26 @@ static int AddExtInfoStat(struct FileEntry *file, unsigned char type, void *data
...
 	FILE *InputFile;
 	int result;
 	unsigned int FileDateStamp;
 	unsigned short FileVersion;
 	if ((InputFile = fopen(path, "rb")) != NULL) {
 		const char* fname = strrchr(path, PATHSEP);
 		if (fname == NULL) fname = path; else fname++;
+        if (upperconv) {
+            strncpy(tbuf, fname, sizeof(tbuf));
+            tbuf[sizeof(tbuf) - 1] = '\0';
```
Comment (fjtrujy) [inline tools/romimg/src/romimg.c:467]: if you initialize with `= {0}` then this line shouldn't be required either

File: tools/romimg/src/platform.h (hunk 5 above)
Comment (israpps) [reply tools/romimg/src/platform.h:17]: windows has the name flipped `MAX_PATH` instead of  `PATH_MAX`

Comment (fjtrujy) [reply tools/romimg/src/platform.h:17]: I'm using `MAX_PATH` in my fork and all the CI/CD was passing.
You are probably speaking about compiling natively, not through MinGW, in theory, we shouldn't care about these platforms...

Comment (israpps) [reply tools/romimg/src/platform.h]: not sure where it is defined then... everything in google says on limits.h but it's not the case

**Review by fjtrujy — APPROVED** [review]

**Pull request merged**